modules/infra/control/vrf_netlink.c

```diff
@@ -0,0 +1,274 @@
+	nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK;
+
+	ifi = mnl_nlmsg_put_extra_header(nlh, sizeof(*ifi));
+	ifi->ifi_family = AF_UNSPEC;
+	ifi->ifi_index = ifindex;
+
+	return netlink_send_req(nlh);
+}
+
+static int create_vrf_and_enslave(const char *vrf_name, uint32_t vrf_table, uint32_t loop_ifindex) {
+	uint32_t vrf_ifindex;
+	int ret;
+
+	ret = link_add_vrf(vrf_name, vrf_table);
+	if (ret < 0)
+		return ret;
+
+	vrf_ifindex = if_nametoindex(vrf_name);
+	if (!vrf_ifindex)
+		return errno_set(ENODEV);
+
+	ret = link_set_master(loop_ifindex, vrf_ifindex);
+	if (ret < 0)
+		return ret;
+
+	ret = link_set_admin_state(vrf_ifindex, true);
+	if (ret < 0)
+		return ret;
+
+	ret = link_set_admin_state(loop_ifindex, true);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static int delete_vrf_and_unslave(const char *vrf_name, uint32_t loop_ifindex) {
+	uint32_t vrf_ifindex;
+	int ret;
+
+	vrf_ifindex = if_nametoindex(vrf_name);
+	if (!vrf_ifindex)
+		return errno_set(ENODEV);
+
+	ret = link_set_master(loop_ifindex, 0);
+	if (ret < 0)
+		return ret;
+
+	ret = link_set_admin_state(loop_ifindex, false);
+	if (ret < 0)
+		return ret;
+
+	return link_del_index(vrf_ifindex);
+}
+
+static int add_del_route(uint32_t ifindex, uint32_t table, bool add) {
+	// nlmsghdr + rtmsg + 2x u32 attrs (RTA_TABLE, RTA_OIF)
+	char buf[NLMSG_SPACE(sizeof(struct rtmsg)) + 2 * NLA_SPACE(sizeof(uint32_t))];
+	struct nlmsghdr *nlh;
+	struct rtmsg *rtm;
+	int ret;
+
+	memset(buf, 0, sizeof(buf));
+	nlh = mnl_nlmsg_put_header(buf);
+	nlh->nlmsg_type = add ? RTM_NEWROUTE : RTM_DELROUTE;
+	nlh->nlmsg_flags = NLM_F_REQUEST | NLM_F_ACK;
+	if (add)
+		nlh->nlmsg_flags |= NLM_F_CREATE | NLM_F_EXCL;
+
+	rtm = mnl_nlmsg_put_extra_header(nlh, sizeof(*rtm));
+	rtm->rtm_family = AF_INET;
+	rtm->rtm_table = RT_TABLE_UNSPEC;
+	rtm->rtm_protocol = RTPROT_BOOT;
+	rtm->rtm_scope = RT_SCOPE_UNIVERSE;
+	rtm->rtm_type = RTN_UNICAST;
+	rtm->rtm_dst_len = 0;
+
+	mnl_attr_put_u32(nlh, RTA_TABLE, table);
+	mnl_attr_put_u32(nlh, RTA_OIF, ifindex);
+
+	ret = netlink_send_req(nlh);
+	if (ret < 0)
+		return ret;
+
+	rtm->rtm_family = AF_INET6;
+	ret = netlink_send_req(nlh);
+	if (ret < 0)
+		return ret;
+
+	return 0;
+}
+
+static void iface_event_handler(uint32_t event, const void *obj) {
+	const struct iface *iface = obj;
+	const char *loop_name;
+	uint32_t loop_ifindex;
+	uint32_t table_id;
+	int ret;
+
+	if (iface->type != GR_IFACE_TYPE_LOOPBACK)
+		return;
+
+	loop_name = loopback_get_tun_name(iface);
+	loop_ifindex = if_nametoindex(loop_name);
+	if (!loop_ifindex) {
+		LOG(WARNING, "%s: %s no ifindex skip vrf sync", __func__, loop_name);
+		return;
+	}
+
+	if (iface->vrf_id)
+		table_id = iface->vrf_id + 1000;
+	else
+		table_id = RT_TABLE_MAIN;
+
+	switch (event) {
+	case GR_EVENT_IFACE_POST_ADD:
+		if (iface->vrf_id) {
+			ret = create_vrf_and_enslave(iface->name, table_id, loop_ifindex);
+			if (ret < 0) {
+				LOG(WARNING,
+				    "%s: create vrf %u for %s failed: %s",
+				    __func__,
```

Comment:
`__func__` is redundant. Already included in `LOG()`.

https://github.com/DPDK/grout/blob/236e7440516d3b7d54f7a37c87d517796d3cda9e/main/gr_log.h#L19